internal/cli/auth/register.go

```diff
@@ -20,52 +20,85 @@ import (
 	"fmt"
 	"os"
 
+	"github.com/AlecAivazis/survey/v2"
 	"github.com/mongodb/mongocli/internal/cli/require"
 	"github.com/mongodb/mongocli/internal/config"
 	"github.com/pkg/browser"
 	"github.com/spf13/cobra"
+	"go.mongodb.org/atlas/auth"
 	atlas "go.mongodb.org/atlas/mongodbatlas"
 )
 
 const accountURI = "https://account.mongodb.com/account/register?fromURI=https://account.mongodb.com/account/connect"
 const govAccountURI = "https://account.mongodbgov.com/account/register?fromURI=https://account.mongodbgov.com/account/connect"
 
+type registerSurvey struct {
+	confirm func(message string, defaultResponse bool) (response bool, err error)
+}
+
 type RegisterOpts struct {
-	login loginOpts
+	login          loginOpts
+	registerSurvey *registerSurvey
+}
+
+var defaultRegisterSurvey = registerSurvey{
+	confirm: func(message string, defaultResponse bool) (response bool, err error) {
+		p := &survey.Confirm{
+			Message: message,
+			Default: defaultResponse,
+		}
+		err = survey.AskOne(p, &response)
+		return response, err
+	},
 }
 
 func (opts *RegisterOpts) registerAndAuthenticate(ctx context.Context) error {
 	// TODO:CLOUDP-121210 - Replace with new request and remove URI override.
-	code, _, err := opts.login.flow.RequestCode(ctx)
-	if err != nil {
-		return err
-	}
+	for {
+		code, _, err := opts.login.flow.RequestCode(ctx)
+		if err != nil {
+			return err
+		}
 
-	if opts.login.isGov {
-		code.VerificationURI = govAccountURI
-	} else {
-		code.VerificationURI = accountURI
-	}
+		if opts.login.isGov {
+			code.VerificationURI = govAccountURI
+		} else {
+			code.VerificationURI = accountURI
+		}
 
-	opts.login.printAuthInstructions(code)
+		opts.login.printAuthInstructions(code)
 
-	if !opts.login.noBrowser {
-		if errBrowser := browser.OpenURL(code.VerificationURI); errBrowser != nil {
-			_, _ = fmt.Fprintf(os.Stderr, "There was an issue opening your browser\n")
+		if !opts.login.noBrowser {
+			if errBrowser := browser.OpenURL(code.VerificationURI); errBrowser != nil {
+				_, _ = fmt.Fprintf(os.Stderr, "There was an issue opening your browser\n")
+			}
 		}
+
+		accessToken, _, err := opts.login.flow.PollToken(ctx, code)
+		if retry, errRetry := opts.shouldRetryRegister(err); errRetry != nil {
+			return errRetry
+		} else if retry {
+			continue
+		}
+
+		if err != nil {
```

Comment:
[q] should this be on line 78?

Reply:
at line 78 we check if `err` is a retriable error and if it is we check if we were able to read the user's answer (`errRetry `).
Here (line 84) we deal with the case when a non-retriable error occurred (e.g. cannot access the api anymore) or the user said he doesn't want to retry anymore (in which case we display the last error: token expired)

Reply:
Thanks, I missed that we're passing error to the retry check.